File: Published/SourceCode/Java/DynamicArrayStack.java

import java.util.Iterator;

class DynamicArrayStack<E> implements Stack<E> {
    private E[] internalArray;   // Internal array containing the stack elements
    private int stackSize;       // Size of stack, and index of the next free slot

    @SuppressWarnings("unchecked")
    public DynamicArrayStack() {
        internalArray = (E[]) new Object[1];
        stackSize = 0;
    }

    public void push(E x) {
        if (stackSize >= internalArray.length) {
            resizeArray(2 * internalArray.length);
        }
        internalArray[stackSize] = x;
        stackSize++;
    }

    public E peek() {
        if (!(stackSize > 0)) throw new IndexOutOfBoundsException("peek from empty stack");
        return internalArray[stackSize-1];
    }

    public E pop() {
        if (!(stackSize > 0)) throw new IndexOutOfBoundsException("pop from empty stack");
        stackSize--;
        E x = internalArray[stackSize];
        internalArray[stackSize] = null;   // For garbage collection
        if (stackSize <= internalArray.length / 3) {
            resizeArray(internalArray.length / 2);
        }
        return x;
    }

    private void resizeArray(int newCapacity) {
        @SuppressWarnings("unchecked")
        E[] newArray = (E[]) new Object[newCapacity];
        for (int i = 0; i < stackSize; i++) {
            newArray[i] = internalArray[i];
        }
        internalArray = newArray;
    }

    public boolean isEmpty() {
        return stackSize == 0;
    }

    public int size() {
        return stackSize;
    }

    public Iterator<E> iterator() {
        return new DynamicArrayStackIterator();
    }

    private class DynamicArrayStackIterator implements Iterator<E> {
        private int index;
        DynamicArrayStackIterator() {
            index = stackSize;
        }
        public boolean hasNext() {
            return index > 0;
        }
        public E next() {
            index--;
            return internalArray[index];
        }
    }


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printList() {
        System.out.print(internalArray.length + " [ ");
        for (int i=size(); i<internalArray.length; i++) System.out.print("- ");
        System.out.print("| ");
        for (E e : this) System.out.print(e + " ");
        System.out.println("] " + size());
    }

    public static void main(String[] args) {
        DynamicArrayStack<String> list = new DynamicArrayStack<>();
        for (int i=0; i<23; i++) {
            list.push(String.valueOf((char)(i+65)));
            if (list.size() % 5 == 0) list._printList();
        }
        list._printList();
        while (!list.isEmpty()) {
            if (list.peek() != list.pop()) throw new IndexOutOfBoundsException("ERROR!!!");
            if (list.size() % 3 == 2) list._printList();
        }
        list._printList();
    }
}

File: Published/SourceCode/Python/DynamicArrayStack.py

from API import Stack

class DynamicArrayStack(Stack):
    _minCapacity = 8            # Minimum capacity of internalArray
    _minLoadFactor = 0.5        # Must be smaller than 1/CapacityMultiplier
    _capacityMultiplier = 1.5   # Factor to grow/shrink the capacity

    def __init__(self):
        self._internalArray = [None] * self._minCapacity   # Internal array containing the stack elements
        self._stackSize = 0                                # Size of stack, and index of the next free slot

    def push(self, x):
        if self._stackSize >= len(self._internalArray):
            self._resizeArray(len(self._internalArray) * self._capacityMultiplier)
        self._internalArray[self._stackSize] = x
        self._stackSize += 1

    def peek(self):
        if not (self._stackSize > 0): raise IndexError("peek from empty stack")
        return self._internalArray[self._stackSize-1]

    def pop(self):
        if not (self._stackSize > 0): raise IndexError("pop from empty stack")
        self._stackSize -= 1
        x = self._internalArray[self._stackSize]
        self._internalArray[self._stackSize] = None   # For garbage collection
        if self._stackSize <= len(self._internalArray) * self._minLoadFactor:
            self._resizeArray(len(self._internalArray) / self._capacityMultiplier)
        return x

    def _resizeArray(self, newCapacity):
        if newCapacity < self._minCapacity: return
        newArray = [None] * int(newCapacity)
        for i in range(self._stackSize):
            newArray[i] = self._internalArray[i]
        self._internalArray = newArray

    def isEmpty(self):
        return self._stackSize == 0

    def size(self):
        return self._stackSize

    def __iter__(self):
        for i in reversed(range(self._stackSize)):
            yield self._internalArray[i]


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printList(l):
    print(len(l._internalArray), "[", "- " * (len(l._internalArray) - l.size()) + "|",
               " ".join(str(e) for e in l), "]", l.size())

if __name__ == '__main__':
    a = DynamicArrayStack()
    for i in range(23):
        a.push(chr(i+65))
        if a.size() % 5 == 0:
            _printList(a)
    _printList(a)
    while not a.isEmpty():
        assert a.peek() == a.pop(), (a,)
        if a.size() % 3 == 2:
            _printList(a)
    _printList(a)
